resize-observer/observe.html

```diff
@@ -966,6 +1009,75 @@
   return helper.start(() => t.remove());
 }
 
+function test20() {
+  // <div id="outer">
+  //   <div id="nested"></div>
+  //   <div id="nested"></div>
+  // </div>
+
+  const size = 50;  // must be an even number
+  const px = v => `${v}px`;
+  const outer = document.createElement('div');
+  // We must pick an odd number of device pixels so that between two `flex: 1 1 auto` children
+  // the only way to find out which of the 2 got the extra pixel is via devicePixelContextBoxSize
+  const outerDevicePixelWidth = Math.round(devicePixelRatio * size + 1);
```

Comment:
Can you explain a bit what you are trying to do here? It seems like you create two children inside the flexbox, and then make sure that 
flexbox.deviceSize == sum(children.deviceSize).

Why is outer sizxe dynamic (depends on devicePixelRation). Why not just make fkr example 99 pixels?

Reply:
Just checking but what does `fkr` mean?

The code is trying to force the browser into a situation that `devicePixelContentBoxSize` was made to solve. It's only purpose for existing is in situations where 2 or more elements get fractional sizes where at the DOM level (elem.clientWidth, elem.getBoundingClientRect, and even ResizeObserverEvent.contentBoxSize[0].inline) all of the values will be equal but what's actually rendered, they are not equal.

So, if I pick "99px" as the parent width and divide by 2 elements I'll get these results on all browsers (*)

```
elem.getBoundingClientRect().width == 49.5
elem.clientWidth = 50;
ResizeObserverEvent.contentBoxSize[0].inlineSize = 49.5
```

But what's actually rendered will will be different depending on the devicePixelRatio (dpr)

On a dpr = 1 device, one element will get 50 pixels, the other 49. Which one? That's what devicePixelContentBoxSize answers.

On a dpr = 2 device, both elements will get 99 device pixels because 99px (CSS pixels) becomes 198 device pixels and you can acutally divide 198 by 2 elements

So, the point of the caluclation above is to try to choose a CSS size that in devicePixels is indivisible by 2 in devicePixels. (or whatever number of children are added). This way you force the browser to chose different sizes for the children and force `devicePixelContentBoxSize` into it's reason for existence (figuring out how the available devicePixels where actually distributed to the children)

So for example on a dpr = 2 device, choosing '99px' pixels would fail to show an issue since it's evenlly divisible by 2 in device pixels. But choosing '99.5px' would provoke the issue for 2 children as 99.5px would be 199 device pixels which is not divisible by 2 so one element is oging to get 99 device pixel and the other 100 device pixels. Again, which one gets the extra pixel is what `devicePixelContentBoxSize` exists to find out.

As for having 2 children vs 47 children, that's just to try enough different values to provoke a bug in implementations.

At the moment, Firefox incorrectly implements `devicePixelContentBoxSize`. There it's just hardcoded to return `entry.contentBoxSize[0].inlineSize * devicePixelRatio` which is incorrect and this update in the test is to point that out.

Reply:
The (*) above is that it's really all browsers except for the rounding issue mentioned in the first comment.

Reply:
> Just checking but what does fkr mean?

It is a misspelling of `for`.

I get the whole devicePixelContentBox, was there when the standard was designed.
I was just checking, and it seems that we already have a test for `devicePixelContentBoxSize`.
It is in `devicepixel.html`. But, that test is a reference test and it passes in FF.

Your test asserts that:
- if you have a flexbox with a row of children
- the total devicePixelContentBoxSize of the children will be equal to the devicePixelContentBoxSize of the flexbox

Is this something we can assert from the standard?
The standard for devicePixelContentBoxSize is
https://drafts.csswg.org/resize-observer/#dom-resizeobserverboxoptions-device-pixel-content-box
It defines devicePixelContentBox as "content box that is mysteriously rounded for optimal painting".

Although what your test asserts is intuitive, I do not think that infer that behavior from  devicePixelContentBoxSize definition. All it says is that devicePixelContentBoxSize is something that looks good when painted. It is a paint concept, not a layout concept. And you are asserting a layout concept.

Is there a real-world issue that you are running into? 

When devicePixelContentBoxSize was originally designed, the use case was `<CANVAS>`. The high DPI canvas drawing hacks needed to know the size of the painted canvas.

Reply:
> Is there a real-world issue that you are running into?

Yes, Firefox returns the wrong values and so I'm not able to render pixel perfect in Firefox.
It arguably returns the wrong values because there's no test so they didn't notice they were doing it wrong (https://bugzilla.mozilla.org/show_bug.cgi?id=1774135)

This test to me is correct. I'm pretty sure the spec says children created as the test creates then must fill their parent perfectly. If you make children in a flexbox with the children set to distribute the space of the parent they must fill the parent.

The test does not check the size of specific children, only that they their sizes sum up to the reported size of their parent. If they don't then the browser clearly as a bug.

The ref test is incomplete and it's impossible to test this case using a ref test since which elements get the extra pixels in undefined.

This seems like a valid and useful test.

> When devicePixelContentBoxSize was originally designed, the use case was <CANVAS>. The high DPI canvas drawing hacks needed to know the size of the painted canvas.

Exactly! A typical use case is an editor with a canvas on one side and a vertical toolbar. Finding out the exact size of the canvas element in this case will fail in Firefox. This test would have caught that issue.

Reply:
> I'm pretty sure the spec says children created as the test creates then must fill their parent perfectly

You are talking about layout dimensions, not paint dimensions. These are 2 different concepts. 
Most of the HTML/CSS spec talks about layout dimensions, not paint dimensions.
Before devicePixelContentBoxSize, there was never a way to observe paint dimensions directly.
That is why that API was invented.

So just because the standard says the layout boxes must fill their parent does not tell
you anything about paint dimensions.

What happens is that internally in the browser, DOM gets laid out and elements get layout
dimensions. Then, when we paint, to make things look better, those dimensions might get
adjusted internally just for painting. What all browsers do is pixel snapping. To avoid making
things look blurry, the align elements on pixel boundaries. How this is done is inconsistent
between browsers. For example, FF snaps borders before layout, while Chrome snaps
borders during paint (there are bugs filed for this).

So, there is no way to write a non-reference test tests devicePixelContentBoxSize size.

An interesting fact about paint pixel snapping is that in Chrome it depends on both element's size 
and location. 

https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/platform/geometry/layout_unit.h;drc=043baafbe3cfbf96887b4f20e57be9f6286e7e39;bpv=1;bpt=1;l=784

The way I used to test our code was to create a hiDPI hack canvas with a size of 100.5 pixels, and then
move its location in 0.1px increments. This would make paint dimensions change between 100 and 101 pixels (while 
physical size remained the same). Without ResizeObserver callback that repainted the canvas, the 101 pixels would look bad.

But, there is nothing in the standard that says that paint size should oscillate between 100 and 101 pixels. I can't think of a way to write a "standard compliant" non-reference test for this. 

The closest thing I can imagine to standards compliant is to write a test that approximates what I used to do to test this: have a 100.5px sized canvas, and move it into left:10px through left:11px in .1 pixel increments, with requestAnimationFrame between each change (because we are testing paint size, not layout). If browser did not report changes in devicePixelContentBoxSize while the element moved, they have most likely not implemented devicePixelContentBoxSize correctly. This would correctly detect today's failures, but different imaging pipelines of the future might fail, even though they are behaving correctly. Or there could be some low-end phone that does something weird....

It looks like FF is aware of the issue. The bug was reported only 18 days ago, and they are thinking about it. I'd expect them to contribute a wpt test case with their fix.

Reply:
> The closest thing I can imagine to standards compliant is to write a test [...]

I could also imagine a test with two superimposed divs, with the same top-left position, whose widths differ by 0.5px (or better: by half a device pixel, i.e. `0.5 / window.devicePixelRatio`).  The foreground div would be the slightly-skinnier one, so that if their pixel-snapped widths end up differing, then you'll see 1px of the background div on the right side.

The test could then use ResizeObserver to measure and compare the `devicePixelContentBox` of these divs.  If it's the same, then we assert that the test is pixel-for-pixel identical to a reference case that **only has the foreground div**.  If it's different, then we assert that the test is not identical to that reference case. (Maybe we could construct a reliable reference case for that not-equal scenario, too, but that might involve additional complexity.)

> It looks like FF is aware of the issue. The bug was reported only 18 days ago, and they are thinking about it. I'd expect them to contribute a wpt test case with their fix.

Indeed we are looking; hopefully we'll have a fix within a month or so.  We would indeed include a WPT test of some sort (possibly along these^ lines, or along the lines of the "should have notified" test that @atotic mentioned), if there isn't already one in the testsuite at that point.

Reply:
(Actually: my superimposed-divs scenario might not be theoretically-pure; I can imagine an engine where adding a small epsilon to its specified width might cause a box to pixel-snap leftwards or rightwards, while still producing the same device-pixel width. Such a scenario would cause my described test to fail, I think.  So: sorry for the maybe-tangent... I'll think more about this as fodder for how-to-test the eventual Firefox patch.)

Reply:
> The closest thing I can imagine to standards compliant is to write a test that approximates what I used to do to test this: have a 100.5px sized canvas, and move it into left:10px through left:11px in .1 pixel increments, with requestAnimationFrame between each change (because we are testing paint size, not layout).

Why is that any more standards complient. AFAIK you've basically told me the spec allows browsers to make elements any size they want, period. So what makes your test case special that the browser isn't just allowed to always paint the div 100 pixels?

> I can't think of a way to write a "standard compliant" non-reference test for this.

I can't think of a way to write a reference test for this.

Also, here's a real-world test case

https://greggman.github.io/doodles/test/resizeobserver-devicepixelcontentbox/

With factional devicePixelRatio it's clear Firefox is not returning the correct size (As above, the easiest way to get a fractional devicePixelRatio is to zoom though you can also use a machine with a fractional dpr or configure one). The issue is subtle, at least on my Mac because the mac is doing it's own scaling and has a hi-dpi, but when the wrong size is returned the canvas (right) will fade darker from left to right. On my PC it's less subtle.

Reply:
>  So what makes your test case special that the browser isn't just allowed to always paint the div 100 pixels?

As I've mentioned, browser is not required to behave in the way I suggested. The fact is, because all browsers implement pixel-snapping based on location/size, my test case should work in what is implemented today.

The difference between mine and your test case is that my case does not require comparing the sum of children's sizes to the size of the parent. The summation part feels like something that might easily break. My case also tests  `devicePixelContentBox` observation firing when layout size does not change, but paint size does, which is the tricky bit to get right.

Reply:
> Also, here's a real-world test case
> https://greggman.github.io/doodles/test/resizeobserver-devicepixelcontentbox/

I like how this test case is using a 2D pattern. All my test cases involved 3D shaders I did
not understand.

Reply:
I'm confused at distinction you're making

Where in the spec is it legal for elements being scaled to fit their parent NOT actually fit their parent? How would any of the web platform work if that was ok? 

Here's a ref test, but again, it's adding up sizes

https://github.com/web-platform-tests/wpt/pull/34682

Reply:
> the spec is it legal for elements being scaled to fit their parent NOT actually fit their parent

There are two different things: layout size, and paint size. They are not always the same.
What you are saying holds true for layout size.

Paint size is left undefined, browsers are free to do whatever they want to make the output look good.

devicePixelContentBox is the only place in the platform where you can observe the paint size.
The only reason this got exposed is to do exactly what you are trying to do, scale canvas
without weird patterns.

Reply:
> devicePixelContentBox is the only place in the platform where you can observe the paint size. The only reason this got exposed is to do exactly what you are trying to do, scale canvas without weird patterns.

And hence my test. I can't stop the patterns without counting on being able to compute the positions. (see the ref test https://github.com/web-platform-tests/wpt/pull/34682). If I can compute the positions then I can make things align. If I can't then what was the point of devicePixelContentBox?

Reply:
Again, for clarification: Let's say I make this HTML

```html
<div id="a"></div>
<div id="b">
  <div id="c"></div>
  <div id="d"></div>
</div>
```

with this CSS

```css
#a, #b {
  width: 99px;
  height: 15px;
}
#b { display: flex; }
#c, #d {
  flex: 1 1 auto;
  height: 15px;
}


#a { background-color: #79fb9a; }
#c { background-color: #ff4949; }
#d { background-color: #2d72d9; }
body { background: #000 }
```

Where the expected result is something like this

![goal](https://user-images.githubusercontent.com/234804/177024230-a798e1cb-3bce-48f3-bad1-237c06a50681.png)

[live](https://jsgist.org/?src=71c24392aa49a7995c492f68116dcc8c)

You're saying that according to the spec, all of these results are 100% valid?

![l-lr](https://user-images.githubusercontent.com/234804/177024268-0a0925d5-89da-4cb7-8863-be79f7c87383.png)

![l-extra](https://user-images.githubusercontent.com/234804/177024270-7ec9af1c-825d-4853-be44-2164f9d0cd1f.png)

![l-lit](https://user-images.githubusercontent.com/234804/177024274-7f38b57e-97a7-4493-9a8f-1c1ae9dc342e.png)

![l-rand](https://user-images.githubusercontent.com/234804/177024299-882d7219-2aee-4915-9552-12b539ba6baa.png)

![l-squish](https://user-images.githubusercontent.com/234804/177024315-5b90fad0-257d-43e1-980f-94d092ed5d00.png)

![l-gap](https://user-images.githubusercontent.com/234804/177024365-bcdeddd5-7efd-44a3-9c3d-c79291a57a3e.png)

![l-leadgap](https://user-images.githubusercontent.com/234804/177024367-65e3eb59-aa18-4bce-a76e-ad6c8c7708ed.png)

Reply:
Aleks is right that the spec does not say how pixel snapping for paint should work, and in that sense some of the variants you list above are "spec compliant". (The specs for painting are quite limited actually, and don't say very much.) The spec also does not say that the edges of the painted pixels need to touch and also need to add up to the width of the container. However, all the browser implementations know that that is the desired outcome. Perhaps we could add normative spec text for it that says "browsers should make pixels add" or similar. Can you file an issue at the CSSWG about this?

Stepping back, I think that what you're ultimately trying to do is test whether the reported `devicePixelContentBoxSize` is exactly the same size as the actual number of device pixels drawn to the screen. And your motivation for doing so is that there is a Firefox bug where it's incorrectly multiplying by `devicePixelRatio` rather than running things through the (more subtle and complicated) pixel snapping routines of the browser.

Is that a correct description? If so, unfortunately I don't think there is a way in WPT to assert that, because you can't read back pixels from the screen with javascript and compare them. The best you could do is I think something like [this test](https://github.com/web-platform-tests/wpt/pull/34682) that you have already written, but technically it's not testing something browsers are required to do. I do think it's reasonable to add it anyway though, but with a caveat note in the test. I also like that it's using canvas to test, which as Aleks mentions is the reason for this feature in the first place.

Re failing tests with DPR 1.33: I agree with Aleks that this looks like a Chromium bug. Can you file one? My first guess is that it's related to the 1/60 resolution limit of CSS pixels in Chromium, but would have to debug further to know for sure.

Reply:
filed chrome bug: https://bugs.chromium.org/p/chromium/issues/detail?id=1345006

filed CSSWG issue: https://github.com/w3c/csswg-drafts/issues/7506 (hope that's the correct place)